libs/wingsdk/src/core/plugin-manager.ts

```diff
@@ -0,0 +1,190 @@
+import { readFileSync, writeFileSync } from "fs";
+import { resolve } from "path";
+import * as vm from "vm";
+import { IConstruct } from "constructs";
+
+/**
+ * Behavioral interface of a plugin hook.
+ */
+export interface ICompilationHook {
+  /**
+   *  name of plugin can be set by plugin as export
+   *
+   * @default - absolute path to plugin file
+   */
+  pluginName: string;
+  /** preSynth phase */
+  preSynth?(app: IConstruct): void;
+  /** postSynth phase */
+  postSynth?(config: any): any;
+  /** validate phase */
+  validate?(config: any): void;
+}
+
+enum PluginPhases {
+  PRE_SYNTH = "preSynth",
+  POST_SYNTH = "postSynth",
+  VALIDATE = "validate",
+}
+
+export interface PluginManagerProps {
+  /**
+   * vm context to load plugins in with
+   *
+   * @default - create a new context with (require, console, exports, process, __dirname)
+   */
+  context?: vm.Context;
+}
+
+/**
+ * Plugin manager is responsible for loading hooks from plugins and
+ * managing their executions during the various phases of the plugin lifecycle.
+ */
+export class PluginManager {
+  private hooks: ICompilationHook[];
+  private context?: vm.Context;
+
+  constructor(plugins: string[], props: PluginManagerProps = {}) {
+    this.hooks = [];
+    this.context = props.context;
+
+    for (const plugin of plugins) {
+      this.add(plugin);
+    }
+  }
+
+  /**
+   * Add plugins to the plugin manager based on the plugin type.
+   *
+   * @param plugin the plugin to add
+   */
+  public add(plugin: string) {
+    // maybe we support other plugin types in the future (e.g. npm modules)
+    if (!plugin.endsWith(".js")) {
+      throw new Error(
+        `Currently only javascript files are supported as plugins. Got: ${plugin}`
+      );
+    }
+    this.addPluginFromFilePath(plugin);
+  }
+
+  /**
+   * Loads hooks from a plugin file and adds them to the list of hooks
+   * that will later be run in the various plugin phases.
+   *
+   * @param filePath absolute file path to the plugin
+   */
+  private addPluginFromFilePath(filePath: string) {
+    const pluginHooks: ICompilationHook = {
+      pluginName: filePath, // can be overwritten by plugin (ex: exports.pluginName = "my-plugin")
+    };
+
+    const pluginDir = resolve(filePath).split("/").slice(0, -1).join("/");
+
+    const context =
+      this.context ??
+      vm.createContext({
+        require,
+        console,
+        exports: pluginHooks,
+        process,
+        __dirname: pluginDir,
+      });
+
+    const pluginCode = readFileSync(resolve(filePath), "utf8");
+    const script = new vm.Script(pluginCode);
+    script.runInNewContext(context);
```

Comment:
I'm not sure but I think `runInContext(context)` might work here since `context` has already been contextified.

Reply:
Yea I kept it for a while incase a custom context was passed then the context does not get contextified. I think that 
 allowing a custom context to be used adds complexity where I dont have a use case for it now. So for now I removed the ability to provide a context and changed this back to just `runInContext`